sortedcontainers/sorteddict.pyi

```diff
@@ -0,0 +1,121 @@
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Generic,
+    Hashable,
+    Iterator,
+    Iterable,
+    ItemsView,
+    KeysView,
+    List,
+    Mapping,
+    Optional,
+    Sequence,
+    Type,
+    TypeVar,
+    Tuple,
+    Union,
+    ValuesView,
+    overload,
+)
+
+_T = TypeVar("_T")
+_S = TypeVar("_S")
+_KT = TypeVar("_KT", bound=Hashable)  # Key type.
+_VT = TypeVar("_VT")  # Value type.
+_KT_co = TypeVar("_KT_co", covariant=True, bound=Hashable)
+_VT_co = TypeVar("_VT_co", covariant=True)
+_SD = TypeVar("_SD", bound=SortedDict)
+_Key = Callable[[_T], Any]
+
+class SortedDict(Dict[_KT, _VT]):
+    @overload
+    def __init__(self, **kwargs: _VT) -> None: ...
+    @overload
+    def __init__(self, __map: Mapping[_KT, _VT], **kwargs: _VT) -> None: ...
+    @overload
+    def __init__(
+        self, __iterable: Iterable[Tuple[_KT, _VT]], **kwargs: _VT
+    ) -> None: ...
+    @overload
+    def __init__(self, __key: _Key[_KT], **kwargs: _VT) -> None: ...
+    @overload
+    def __init__(
+        self, __key: _Key[_KT], __map: Mapping[_KT, _VT], **kwargs: _VT
+    ) -> None: ...
+    @overload
+    def __init__(
+        self,
+        __key: _Key[_KT],
+        __iterable: Iterable[Tuple[_KT, _VT]],
+        **kwargs: _VT
+    ) -> None: ...
+    @property
+    def key(self) -> Optional[_Key[_KT]]: ...
+    @property
+    def iloc(self) -> SortedKeysView[_KT]: ...
+    def clear(self) -> None: ...
+    def __delitem__(self, key: _KT) -> None: ...
+    def __iter__(self) -> Iterator[_KT]: ...
+    def __reversed__(self) -> Iterator[_KT]: ...
+    def __setitem__(self, key: _KT, value: _VT) -> None: ...
+    def _setitem(self, key: _KT, value: _VT) -> None: ...
+    def copy(self: _SD) -> _SD: ...
+    def __copy__(self: _SD) -> _SD: ...
+    @classmethod
+    @overload
+    def fromkeys(cls, seq: Iterable[_T]) -> SortedDict[_T, None]: ...
+    @classmethod
+    @overload
+    def fromkeys(cls, seq: Iterable[_T], value: _S) -> SortedDict[_T, _S]: ...
+    def keys(self) -> SortedKeysView[_KT]: ...
+    def items(self) -> SortedItemsView[_KT, _VT]: ...
+    def values(self) -> SortedValuesView[_VT]: ...
+    @overload
+    def pop(self, key: _KT) -> _VT: ...
+    @overload
+    def pop(self, key: _KT, default: _T = ...) -> Union[_VT, _T]: ...
+    def popitem(self, index: int = ...) -> Tuple[_KT, _VT]: ...
+    def peekitem(self, index: int = ...) -> Tuple[_KT, _VT]: ...
+    def setdefault(self, key: _KT, default: Optional[_VT] = ...) -> _VT: ...
```

Comment:
Why specify `default` here and not use an overload as with `fromkeys`?

Reply:
With `fromkeys()`, the return value is dependent upon whether `value` is specified or not. I went and checked the implementation of `setdefault()` and I can see that this typing is wrong and it should actually be:

```python
    def setdefault(self, key: _KT, default: Optional[_VT] = ...) -> Optional[_VT]: ...
```

@althonos

Reply:
Hmmm... the stdlib has the following:

```python
    def setdefault(self, k: _KT, default: Optional[_VT] = ...) -> _VT: ...
```

However, I can do the following:

```python
d = SortedDict()
v = d.setdefault('a') # v is None
v = d.setdefault('a', 1) # v is still None
```

I'm going to ask in the mypy gitter room, but returning `_VT` seems wrong based on the code above. I'll let you know what I find out.

Reply:
@bryanforbes : my bad, this would be a thing for `dict.get`, but not for `dict.setdefault`, you're right.

Reply:
@althonos From my conversation in the Gitter room, we should leave this the same as what the stdlib has and I'm going to file an issue to discuss it on typeshed. The `Optional[_VT]` return type might produce a lot of false-positives.